src/evm_transition_tool/transition_tool.py

```diff
@@ -0,0 +1,251 @@
+
+    # Abstract methods that each tool must implement
+
+    @abstractmethod
+    def __init__(
+        self,
+        *,
+        binary: Optional[Path] = None,
+        trace: bool = False,
+    ):
+        """
+        Abstract initialization method that all subclasses must implement.
+        """
+        if binary is None:
+            binary = self.default_binary
+        else:
+            # improve behavior of which by resolving the path: ~/relative paths don't work
+            resolved_path = Path(os.path.expanduser(binary)).resolve()
+            if resolved_path.exists():
+                binary = resolved_path
+        binary = shutil.which(binary)  # type: ignore
+        if not binary:
+            raise TransitionToolNotFoundInPath(binary=binary)
+        self.binary = Path(binary)
+        self.trace = trace
+
+    def __init_subclass__(cls):
+        """
+        Registers all subclasses of TransitionTool as possible tools.
+        """
+        TransitionTool.register_tool(cls)
+
+    @classmethod
+    def register_tool(cls, tool_subclass: Type["TransitionTool"]):
+        """
+        Registers a given subclass as tool option.
+        """
+        cls.registered_tools.append(tool_subclass)
+
+    @classmethod
+    def set_default_tool(cls, tool_subclass: Type["TransitionTool"]):
+        """
+        Registers the default tool subclass.
+        """
+        cls.default_tool = tool_subclass
+
+    @classmethod
+    def from_binary_path(cls, *, binary_path: Optional[Path], **kwargs) -> "TransitionTool":
+        """
+        Instantiates the appropriate TransitionTool subclass derived from the
+        tool's binary path.
+        """
+        assert cls.default_tool is not None, "default transition tool was never set"
+
+        if binary_path is None:
+            return cls.default_tool(binary=binary_path, **kwargs)
+
+        resolved_path = Path(os.path.expanduser(binary_path)).resolve()
+        if resolved_path.exists():
+            binary = resolved_path
+        binary = shutil.which(binary_path)  # type: ignore
+
+        if binary is not None:
+            binary = Path(binary)
+            # Group the tools by version flag, so we only have to call the tool once for all the
+            # classes that share the same version flag
+            for version_flag, subclasses in groupby(
+                cls.registered_tools, key=lambda x: x.version_flag
+            ):
+                try:
+                    binary_output = os.popen(f"{binary} {version_flag}").read()
+                except Exception:
+                    # If the tool doesn't support the version flag,
+                    # we'll get an non-zero exit code.
+                    continue
+                for subclass in subclasses:
+                    if subclass.detect_binary(binary_output):
+                        return subclass(binary=binary, **kwargs)
+
+        raise UnknownTransitionTool(f"Unknown transition tool binary: {binary_path}")
```

Comment:
I think this is the appropriate error handling:
```suggestion
                raise UnknownTransitionTool(f"Unknown transition tool binary: {binary_path}")

        raise TransitionToolNotFoundInPath(binary=binary)
```